### PR TITLE
BUG: Fixing regression in DataFrame.all and DataFrame.any with bool_only=True

### LINKED ISSUE #25101: all/any ignoring `bool_only` parameter
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
import pandas as pd
print(pd.DataFrame({"col1": [1,2,3], "col2": [4,5,6], "col3": [None, None, None]}).all(bool_only=True))
print(pd.DataFrame({"col1": [1,2,3], "col2": [4,5,6], "col3": [None, None, None], "col4":[False, False, True]}).all(bool_only=True))
```
#### Problem description

`DataFrame.all` and `DataFrame.any` are both ignoring `bool_only` parameter. It does not seem to matter if there are booleans in the input or not.

This was a regression in 0.24.0. On 0.23.4 it was working, expected output from code sample below.

#### Expected Output

Expected output is:
```python
Series([], dtype: bool)

col4    False
dtype: bool
```

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
```
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.0
pytest: 3.8.2
pip: 18.0
setuptools: 40.4.3
Cython: 0.29
numpy: 1.15.0
scipy: 1.2.0
pyarrow: 0.12.0-RAY
xarray: None
IPython: 7.0.1
sphinx: 1.8.1
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: 0.4.0
matplotlib: 3.0.0
openpyxl: 2.5.8
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml.etree: 4.2.5
bs4: None
html5lib: None
sqlalchemy: 1.2.17
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: 0.6.1
pandas_datareader: None
gcsfs: None
```
</details>

### PR DESCRIPTION
**Pull request by devin-petersohn**
* `bool_only` parameter is supported again
* Commit 36ab8c95a41bbadc286a5a520883b149c86931bd created this regression
  due to a bug in all/any (pandas-dev/pandas#23070)
* Reverted the regression and fixed the bug with a condition
* Added tests for `bool_only` parameter

- [x] closes #25101 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.1.rst [hunk 1]
```diff
@@ -58,6 +58,7 @@ Fixed Regressions
 - Fixed regression in :func:`merge` when merging an empty ``DataFrame`` with multiple timezone-aware columns on one of the timezone-aware columns (:issue:`25014`).
 - Fixed regression in :meth:`Series.rename_axis` and :meth:`DataFrame.rename_axis` where passing ``None`` failed to remove the axis name (:issue:`25034`)
 - Fixed regression in :func:`to_timedelta` with `box=False` incorrectly returning a ``datetime64`` object instead of a ``timedelta64`` object (:issue:`24961`)
+- Fixed regression in :meth:`DataFrame.all` and :meth:`DataFrame.any` where ``bool_only`` as ``True`` was ignored (:issue:`25101`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.1.rst]: move to 0.24.2

Comment (jreback) [reply doc/source/whatsnew/v0.24.1.rst]: say ``bool_only=True``

File: pandas/tests/frame/test_analytics.py [hunk 2]
```diff
@@ -1442,6 +1442,26 @@ def test_any_datetime(self):
...
+    def test_any_all_bool_only(self):
+
+        # GH 25101
+        df = DataFrame({"col1": [1, 2, 3],
+                        "col2": [4, 5, 6],
+                        "col3": [None, None, None]})
+
+        result = df.all(bool_only=True)
+        expected = Series(dtype=np.bool)
+        tm.assert_equal(result, expected)
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: use assert_series_equal

File: pandas/tests/frame/test_analytics.py [hunk 3]
```diff
@@ -1442,6 +1442,26 @@ def test_any_datetime(self):
         expected = Series([True, True, True, False])
         tm.assert_series_equal(result, expected)
 
+    def test_any_all_bool_only(self):
+
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: is this near the other original test?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_analytics.py (hunk 3 above)
Comment (devin-petersohn) [reply pandas/tests/frame/test_analytics.py]: Yes, it is directly below the test in the previous commit that introduced this bug.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. we are waiting on merging 0.24.2 for a bit.

**Comment by jreback** [comment]
thanks @devin-petersohn

**Pull request merged**